kyo-core/shared/src/main/scala/kyo/StreamCoreExtensions.scala

```diff
@@ -30,6 +30,87 @@ object StreamCoreExtensions:
         Abort.run(emit).unit
     end emitMaybeElementsFromChannel
 
+    sealed trait StreamHub[A, E]:
+        def subscribe(using Frame): Stream[A, Abort[E] & Async] < (Resource & Async)
+
+    private class StreamHubImpl[A, E](
+        hub: Hub[Result.Partial[E, Maybe[Chunk[A]]]],
+        streamStatus: AtomicRef.Unsafe[Maybe[Result.Partial[E, Unit]]],
+        latch: Latch
+    )(
+        using
+        Tag[Emit[Chunk[A]]],
+        Tag[Emit[Chunk[Chunk[A]]]]
+    ) extends StreamHub[A, E]:
+        private def emit(listener: Hub.Listener[Result.Partial[E, Maybe[Chunk[A]]]])(using Frame) =
+            listener
+                .stream(1)
+                .collectWhile[Chunk[A], Abort[E]] {
+                    case Result.Success(maybeChunk) => maybeChunk
+                    case Result.Failure(e)          => Abort.fail(e)
+                }
+                .mapChunk(v => v.flattenChunk)
+                .emit
+        end emit
+
+        private def emitWithStatus(listener: Hub.Listener[Result.Partial[E, Maybe[Chunk[A]]]])(using Frame) =
+            // Ensure the end-of-stream signal is propagated to new listeners
+            IO.Unsafe(streamStatus.get()).map:
+                case Present(Result.Success(_)) =>
+                    Abort.run[Closed](hub.put(Result.Success(Absent))).andThen(emit(listener))
+                case Present(Result.Failure(e)) =>
+                    Abort.run[Closed](hub.put(Result.Failure(e))).andThen(emit(listener))
+                case Absent =>
+                    emit(listener)
+
+        def subscribe(using Frame): Stream[A, Abort[E] & Async] < (Resource & Async) =
+            Abort.runPartial[Closed](hub.listen).map:
+                case Result.Success(listener) =>
+                    Stream:
+                        latch.release.andThen:
+                            Abort.run[Closed](hub.empty).map: hubIsEmptyResult =>
+                                // If the hub is not empty post subscription, we don't need to check status
+                                if hubIsEmptyResult.getOrElse(true) then
+                                    emitWithStatus(listener)
+                                else emit(listener)
+
+                case Result.Failure(e) => Abort.panic(e)
+
+        def consume[S](stream: Stream[A, Abort[E] & S & Async])(
+            using
+            i1: Isolate.Contextual[S, IO],
+            i2: Isolate.Stateful[S, Abort[E] & Async],
+            t1: SafeClassTag[E],
+            t2: Tag[Emit[Chunk[A]]],
+            fr: Frame
+        ): Unit < (Async & S & Resource) =
+            Resource.acquireRelease(Async.run {
+                Abort.run[E](
```

Comment:
It seems this could be `Abort.run[E | Close]`?

Reply:
I only want to propagate the `E` failures to the output stream, so it's convenient to just handle them separately